File: async_trend_UI/lib/trend-data.ts
import { readFile } from "node:fs/promises";
import path from "node:path";
import type { TrendPayload } from "@/lib/types";

function isTrendPayload(value: unknown): value is TrendPayload {
  if (!value || typeof value !== "object") return false;
  const payload = value as TrendPayload;
  return (
    Array.isArray(payload.trend_dates) &&
    Array.isArray(payload.dates_wise_single) &&
    Array.isArray(payload.dates_wise_summary) &&
    payload.dates_wise_table !== null &&
    typeof payload.dates_wise_table === "object"
  );
}

function unwrapTrendPayload(value: unknown): TrendPayload | null {
  if (isTrendPayload(value)) return value;
  if (value && typeof value === "object" && "payload" in value) {
    const candidate = (value as { payload?: unknown }).payload;
    if (isTrendPayload(candidate)) return candidate;
  }
  return null;
}

export async function loadTrendPayload(): Promise<{ source: string; payload: TrendPayload }> {
  const failures: string[] = [];
  const remoteUrl = process.env.TREND_DATA_URL?.trim();
  if (remoteUrl) {
    try {
      const res = await fetch(remoteUrl, { cache: "no-store" });
      if (!res.ok) {
        failures.push(`TREND_DATA_URL HTTP ${res.status} ${res.statusText}`);
      } else {
        const data = (await res.json()) as unknown;
        const payload = unwrapTrendPayload(data);
        if (payload) {
          return { source: "TREND_DATA_URL", payload };
        }
        failures.push("TREND_DATA_URL returned invalid trend payload shape");
      }
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown fetch error";
      failures.push(`TREND_DATA_URL fetch failed: ${message}`);
    }
  }

  const envPath = process.env.TREND_JSON_PATH?.trim();
  const candidatePaths = [
    envPath
      ? path.isAbsolute(envPath)
        ? envPath
        : path.join(process.cwd(), envPath)
      : null,
    path.join(process.cwd(), "public", "trend_tables.json")
  ].filter((value): value is string => Boolean(value));

  for (const resolvedPath of candidatePaths) {
    try {
      const raw = await readFile(resolvedPath, "utf-8");
      const data = JSON.parse(raw) as unknown;
      const payload = unwrapTrendPayload(data);
      if (!payload) {
        failures.push(`Invalid trend payload at ${resolvedPath}`);
        continue;
      }
      return { source: resolvedPath, payload };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown file error";
      failures.push(`${resolvedPath}: ${message}`);
    }
  }

  throw new Error(
    `Unable to load trend data. ${failures.join(" | ")}. Configure TREND_DATA_URL or TREND_JSON_PATH, or provide public/trend_tables.json.`
  );
}

File: upstox_tools/trend_data.py
from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any, Dict, Tuple

import requests


TrendPayload = Dict[str, Any]


def _candidate_paths() -> list[Path]:
    generated = Path(
        os.getenv("GENERATED_TREND_JSON_PATH", "").strip() or "data/trend_tables.json"
    )
    env_path = os.getenv("TREND_JSON_PATH", "").strip()
    fallback = Path("async_trend_UI") / "public" / "trend_tables.json"
    paths: list[Path] = [generated]
    if env_path:
        paths.append(Path(env_path))
    paths.append(fallback)
    deduped: list[Path] = []
    seen: set[str] = set()
    for path in paths:
        resolved = str(path.resolve()) if path.exists() else str(path)
        if resolved in seen:
            continue
        seen.add(resolved)
        deduped.append(path)
    return deduped


def _unwrap_payload(value: Any) -> TrendPayload | None:
    if not isinstance(value, dict):
        return None
    payload = value.get("payload", value)
    if not isinstance(payload, dict):
        return None
    required = (
        "trend_dates",
        "dates_wise_single",
        "dates_wise_summary",
        "dates_wise_table",
    )
    if not all(key in payload for key in required):
        return None
    if not isinstance(payload["dates_wise_table"], dict):
        return None
    return payload


def load_trend_payload() -> Tuple[str, TrendPayload]:
    failures: list[str] = []

    for path in _candidate_paths():
        try:
            raw = path.read_text(encoding="utf-8")
        except OSError as exc:
            failures.append(f"{path}: {exc}")
            continue
        try:
            data = json.loads(raw)
        except json.JSONDecodeError as exc:
            failures.append(f"{path}: {exc}")
            continue
        payload = _unwrap_payload(data)
        if payload is not None:
            return (str(path), payload)
        failures.append(f"{path}: invalid trend payload shape")

    remote_url = os.getenv("TREND_DATA_URL", "").strip()
    if remote_url:
        try:
            response = requests.get(remote_url, timeout=20)
            response.raise_for_status()
            payload = _unwrap_payload(response.json())
            if payload is not None:
                return ("TREND_DATA_URL", payload)
            failures.append("TREND_DATA_URL: invalid trend payload shape")
        except requests.RequestException as exc:
            failures.append(f"TREND_DATA_URL: {exc}")

    joined = " | ".join(failures) if failures else "No trend sources configured."
    raise RuntimeError(f"Unable to load trend data. {joined}")

